src/main/scala/is/hail/utils/richUtils/RichCodeMemoryBuffer.scala

```diff
@@ -83,18 +83,22 @@ class RichCodeRegion(val region: Code[Region]) extends AnyVal {
     region.invoke[Long, Long, Boolean]("loadBit", byteOff, bitOff)
   }
 
-  def loadPrimitive(typ: Type): Code[Long] => Code[_] = typ match {
+  def loadPrimitive(typ: Type): Code[Long] => Code[_] = typ.fundamentalType match {
     case _: TBoolean =>
-      this.loadBoolean(_)
+      loadBoolean
     case _: TInt32 =>
-      this.loadInt(_)
+      loadInt
     case _: TInt64 =>
-      this.loadLong(_)
+      loadLong
     case _: TFloat32 =>
-      this.loadFloat(_)
+      loadFloat
     case _: TFloat64 =>
-      this.loadDouble(_)
-    case _ =>
+      loadDouble
+    case _: TArray =>
+      loadAddress
+    case _: TBinary =>
+      loadAddress
+    case _: TStruct =>
```

Comment:
I find catch all code like the old `case _ => off => off` to be somewhat dangerous (esp. if it is supposed to handle a few cases.  By enumerating them, (1) you have to think through the cases, (2) you check that verifies you did, and (3) by writing it out explicitly you make potentially make implicit errors in your assumptions explicit.)

Reply:
👍